[feat] engine: implementation of ask.com

Austin-Olacsi opened the pull request:
## What does this PR do?

adds ask.com search engine

## Why is this change important?

the results are similar to other search engines, but i was not able to find where they get results from.

## How to test this PR locally?

searxng !ask

## Author's checklist

<!-- additional notes for reviewers -->

## Related issues

<!--
Closes #234
-->

Austin-Olacsi committed:
[feat] engine: implementation of ask.com

return42 commented on searx/engines/ask.py:
```diff
@@ -0,0 +1,56 @@
+    text = html.fromstring(resp.text).text_content()
+    urls_match = re.findall(r'"url":"(.*?)"', text)
+    titles_match = re.findall(r'"title":"(.*?)"', text)[3:]
+    content_match = re.findall(r'"abstract":"(.*?)"', text)
```
grep a regexp in a XML (HTML) content is fragile (consider the regexp pattern is in the result itself / or in the search term) .. like in all the other engines we should use xpath selectors.

A good starting point are the xpath selectors I have stolen from @allendema's PR https://github.com/searx/searx/pull/2898 / A XPATH engine that we have never merged into SearXNG (was merged into the old project searx but not here in this project SearXNG)

untested:

```python

    title_xpath = '//a[@class="PartialSearchResults-item-title-link result-link"]'
    url_xpath = '//a[@class="PartialSearchResults-item-title-link result-link"]/@href'
    content_xpath = '//p[@class="PartialSearchResults-item-abstract"]'
    suggestion_xpath = '//span[@class="PartialRelatedSearch-item-link-text"]'
```

return42 committed:
[mod] engine ask.com - parse JS result to JSON

Parse the result list from ask.com given in the variable named
window.MESON.initialState::

    <script nonce="..">
        window.MESON = window.MESON || {};
        window.MESON.initialState = {"siteConfig": ...
          ...}};
        window.MESON.loadedLang = "en";
    </script>

The result list is in field::

    json_resp['search']['webResults']['results']

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

Bnyro commented on searx/engines/ask.py:
```diff
@@ -0,0 +1,78 @@
+    dom = html.fromstring(resp.text)
+    script = utils.eval_xpath_getindex(dom, '//script', 0, default=None).text
+
+    pos = script.index(start_tag) + len(start_tag) - 1
+    script = script[pos:]
+    pos = script.index(end_tag) + len(end_tag) - 1
+    script = script[:pos]
```
```suggestion
    start_pos = script.index(start_tag) + len(start_tag) - 1
    end_pos = script.index(end_tag) + len(end_tag) - 1
    script = script[start_pos:end_pos]
```

Bnyro replied:
Or wait, the previous way probably has a slightly better performance because the second `index` call has to read less characters?

return42 replied:
> call has to read less characters?

My version asserts that the `end_tag` is not taken from a string before `start_tag` / assume start_tag is `= {`  end_tag is `};` .. then in your version the first `end_tag` is in the first line but the `start_tag` is in the second line 

```
 window.MESON = window.MESON || {};
 window.MESON.initialState = {"siteConfig": ...
  ...}};
```

Bnyro replied:
Oh, that makes sense :+1:

Bnyro commented on searx/engines/ask.py:
```diff
@@ -0,0 +1,78 @@
+        for field in ['category_l1', 'catsy']:
+            value = item.get(field)
+            if value:
+                metadata.append(value)
```
```suggestion
        fields = ['category_l1', 'catsy']
        metadata = [item.get(field) for field in fields if item.get(field)] 
```

return42 replied:
:+1: --> [done](https://github.com/searxng/searxng/compare/11c2d30b456064c22b29d0a47152a87014649bef..b26e76a1e10056fcfc56f8ffbf1902d97f2ff59f)
